instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraDatabaseClientTracer.java

```diff
@@ -49,14 +56,67 @@ protected InetSocketAddress peerAddress(CqlSession cqlSession) {
     return null;
   }
 
+  @Override
+  protected Span onConnection(Span span, CqlSession cqlSession) {
+    span = super.onConnection(span, cqlSession);
+    DriverExecutionProfile config = cqlSession.getContext().getConfig().getDefaultProfile();
+    // may be overwritten by statement, but take the default for now
+    int pageSize = config.getInt(DefaultDriverOption.REQUEST_PAGE_SIZE);
+    if (pageSize > 0) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_PAGE_SIZE, pageSize);
+    }
+    // may be overwritten by statement, but take the default for now
+    span.setAttribute(
+        SemanticAttributes.DB_CASSANDRA_CONSISTENCY_LEVEL,
+        config.getString(DefaultDriverOption.REQUEST_CONSISTENCY));
+    return span;
+  }
+
   public void onResponse(Context context, ExecutionInfo executionInfo) {
+    Span span = Span.fromContext(context);
     Node coordinator = executionInfo.getCoordinator();
     if (coordinator != null) {
       SocketAddress socketAddress = coordinator.getEndPoint().resolve();
       if (socketAddress instanceof InetSocketAddress) {
-        Span span = Span.fromContext(context);
         NetPeerUtils.INSTANCE.setNetPeer(span, ((InetSocketAddress) socketAddress));
       }
+      if (coordinator.getDatacenter() != null) {
+        span.setAttribute(
+            SemanticAttributes.DB_CASSANDRA_COORDINATOR_DC, coordinator.getDatacenter());
+      }
+      if (coordinator.getHostId() != null) {
+        span.setAttribute(
+            SemanticAttributes.DB_CASSANDRA_COORDINATOR_ID, coordinator.getHostId().toString());
+      }
     }
+    span.setAttribute(
+        SemanticAttributes.DB_CASSANDRA_SPECULATIVE_EXECUTION_COUNT,
+        executionInfo.getSpeculativeExecutionCount());
+
+    Statement<?> statement = executionInfo.getStatement();
+    // override connection default if present
+    if (statement.getConsistencyLevel() != null) {
+      span.setAttribute(
+          SemanticAttributes.DB_CASSANDRA_CONSISTENCY_LEVEL,
+          statement.getConsistencyLevel().name());
+    }
+    // override connection default if present
+    if (statement.getPageSize() > 0) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_PAGE_SIZE, statement.getPageSize());
+    }
+    span.setAttribute(SemanticAttributes.DB_CASSANDRA_IDEMPOTENCE, isIdempotent(statement));
```

Comment:
statements are not idempotent by default. I can't think of a situation where it would be idempotent but not tagged as such. by default it isn't specified though so null should be treated as false. for example, I just tested out running an insert statement which is not idempotent. it had a null idempotent field.

Reply:
the javadoc for isIdempotent() says it returns `null to use the default value defined in the configuration` and references DefaultDriverOption.REQUEST_DEFAULT_IDEMPOTENCE

Reply:
hmm, that's a good catch. I will add it into the same set as the other default driver options.